### PR TITLE
Preliminary Sema and AST support for subclass existentials (SE-0156)

### PR DESCRIPTION
**Pull request by slavapestov**
Work in progress.

TODO:
- If a class C conforms to P, then `C & P` should be the same type as `C`. While we can convert back and forth, they currently appear as distinct types for purposes of overloading, etc.
- If a class C is a subclass of D, the proposal says `C & D` should be equivalent to `D`; for now, it is an error. `C & C` is equivalent to C, though.
- Type alias lookup on a subclass existential is not quite right if the typealias is a member of a protocol that the superclass conforms to concretely.
- Protocol inheritance clauses cannot contain superclasses yet.
- AnyObject is not yet lowered away completely.
- Mangling, SILGen, IRGen, serialization, runtime support.
- Importing Objective-C subclass existentials.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: include/swift/AST/ExistentialLayout.h [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+  ExistentialLayout(ProtocolType *type);
+  ExistentialLayout(ProtocolCompositionType *type);
+
+  /// The superclass constraint, if any.
+  Type superclass;
+
+  /// Whether the existential requires a class, either via an explicit
+  /// '& AnyObject' member or because of a superclass or protocol constraint.
+  bool requiresClass : 1;
```
Comment (DougGregor) [inline include/swift/AST/ExistentialLayout.h]: Should this more generally have a `LayoutConstraintKind` ?

File: lib/AST/Type.cpp [hunk 2]
```diff
@@ -217,53 +218,111 @@ void TypeBase::getAnyExistentialTypeProtocols(
...
+
+  // FIXME: Eventually, there will be a requiresClass() bit in
+  // the ProtocolCompositionType
+  requiresClass = false;
+  requiresClassImplied = false;
+  containsNonObjCProtocol = false;
+
+  auto members = type->getProtocols();
+  if (!members.empty() &&
+      isa<ClassDecl>(members[0]->getAnyNominal())) {
```
Comment (DougGregor) [inline lib/AST/Type.cpp]: `getClassOrBoundGenericClass()` would be slightly more direct.

File: lib/AST/Decl.cpp [hunk 3]
```diff
@@ -2727,16 +2727,16 @@ ProtocolDecl::getInheritedProtocols() const {
...
-          SmallVector<ProtocolDecl *, 4> protocols;
-          type->getExistentialTypeProtocols(protocols);
-          for (auto proto : protocols) {
-            if (known.insert(proto).second)
-              result.push_back(proto);
-          }
+        // Only protocols can appear in the inheritance clause
+        // of a protocol -- anything else should get diagnosed
+        // elsewhere.
+        if (auto *protoTy = type->getAs<ProtocolType>()) {
```
Comment (DougGregor) [inline lib/AST/Decl.cpp]: Wait, really? I thought we could have typealiases of protocol compositions, as well as superclass types, in this list?

File: lib/AST/ASTContext.cpp [hunk 4]
```diff
@@ -2803,16 +2803,19 @@ ProtocolCompositionType::build(const ASTContext &C, ArrayRef<Type> Protocols) {
...
+  RecursiveTypeProperties properties;
   for (Type t : Protocols) {
     if (!t->isCanonical())
       isCanonical = false;
+    properties |= t->getRecursiveProperties();
   }
 
   // Create a new protocol composition type.
   ProtocolCompositionType *New
     = new (C, AllocationArena::Permanent)
```
Comment (DougGregor) [inline lib/AST/ASTContext.cpp]: Wrong arena! You'll need to compute the arena based on the recursive properties.

File: lib/AST/GenericSignatureBuilder.cpp [hunk 5]
```diff
@@ -2507,12 +2508,22 @@ bool GenericSignatureBuilder::addTypeRequirement(
...
-    for (auto proto : protocols) {
-      if (addConformanceRequirement(subjectPA, proto, resolvedSource,
+
+    auto layout = constraintType->getExistentialLayout();
+
+    assert((!layout.requiresClass || layout.requiresClassImplied) &&
+           "explicit AnyObject not yet supported");
+
+    if (layout.superclass) {
+      if (addSuperclassRequirementDirect(subjectPA, layout.superclass,
```
Comment (DougGregor) [inline lib/AST/GenericSignatureBuilder.cpp]: Thanks for fixing me FIXME.

File: lib/AST/Type.cpp (hunk 2 above)
Comment (slavapestov) [inline lib/AST/Type.cpp]: That doesn't support UnboundGenericType, which is possible here.

File: lib/AST/Decl.cpp (hunk 3 above)
Comment (slavapestov) [inline lib/AST/Decl.cpp]: I'll add more tests and see what we support exactly.

File: lib/AST/ASTContext.cpp (hunk 4 above)
Comment (slavapestov) [inline lib/AST/ASTContext.cpp]: Good point, thanks!

Comment (slavapestov) [inline lib/AST/ASTContext.cpp]: I just wanted to give @jrose-apple another fun memory smasher bug to track down :)

File: lib/AST/Type.cpp (hunk 2 above)
Comment (DougGregor) [inline lib/AST/Type.cpp]: Ah, okay!

File: lib/AST/ASTContext.cpp (hunk 4 above)
Comment (DougGregor) [inline lib/AST/ASTContext.cpp]: Just think of how much time we saved with this one little bit of code review.

Comment (jrose-apple) [inline lib/AST/ASTContext.cpp]: \*shudder\*

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Pull request merged**